grammar.js

```diff
@@ -37,25 +44,36 @@ module.exports = grammar({
         $.proc,
         $.char,
         $.symbol,
+        $.instance_variable,
+        $.class_variable,
+        $.self,
         $.identifier,
         $.constant,
+        $.comment,
       ),
 
     string: $ => choice($.quoted_string, $.percent_string),
 
-    chained_string: $ => seq($.string, repeat1($.string)),
-
     quoted_string: $ => seq('"', repeat(/[^"]/), '"'),
 
-    // TODO: redo this lazy implementation
+    // TODO: Allow nested brackets
     percent_string: $ =>
-      seq(
-        '%',
-        choice('(', '[', '{', '<', '|'),
-        repeat(/[^)\]}>|]/),
-        choice(')', ']', '}', '>', '|'),
+      choice(
+        ...bracket_pairs.map(([start, end, escaped_end]) =>
+          seq(
+            '%',
+            optional('q'),
+            start,
+            repeat(
+              new RegExp(`[^${escaped_end}]`),
+            ),
+            end,
+          )
+        ),
       ),
 
+    regex_literal: $ => seq('/', repeat(/[^\/]|\\\//), '/'),
```

Comment:
This doesn't work for expressions following a regex literal (e.g. comments). How do the other implementations do it?

Reply:
I think this is more of a problem of how it's organized currently. I think we should differentiate between statements (which can only exist on a newline - class defs, module defs, method defs) and expressions like strings, literals, binary operations, etc. I've started work on this on a different branch, didn't want to do too much in a PR at once.

Reply:
Ok that's fine, should be addressed in a future PR. 👍